src/wallet.rs

```diff
@@ -0,0 +1,709 @@
+    /// The encryption only happens once and as such, there is conceptually only one message and we are not "reusing" the nonce which would be insecure.
+    pub fn encrypted_xprv_key(&self) -> Result<Vec<u8>> {
+        let cipher = Aes256GcmSiv::new(GenericArray::from_slice(&self.encryption_key));
+        let xprv = &self.xprv.to_string(Prefix::XPRV);
+        let enc_sk = cipher
+            .encrypt(
+                GenericArray::from_slice(SECRET_KEY_ENCRYPTION_NONCE),
+                xprv.as_bytes(),
+            )
+            .context("failed to encrypt secret key")?;
+
+        Ok(enc_sk)
+    }
+
+    /// Derive the encryption key from the wallet's password and a salt.
+    ///
+    /// # Choice of salt
+    ///
+    /// The salt of HKDF can be public or secret and while it can operate without a salt, it is better to pass a salt value [0].
+    ///
+    /// # Choice of ikm
+    ///
+    /// The user's password is our input key material. The stronger the password, the better the resulting encryption key.
+    ///
+    /// # Choice of info
+    ///
+    /// HKDF can operate without `info`, however, it is useful to "tag" the derived key with its usage.
+    /// In our case, we use the encryption key to encrypt the secret key and as such, tag it with `b"ENCRYPTION_KEY"`.
+    ///
+    /// [0]: https://tools.ietf.org/html/rfc5869#section-3.1
+    fn derive_encryption_key(password: &str, salt: &[u8]) -> Result<[u8; 32]> {
+        let h = Hkdf::<Sha256>::new(Some(salt), password.as_bytes());
+        let mut enc_key = [0u8; 32];
+        h.expand(b"ENCRYPTION_KEY", &mut enc_key)
+            .context("failed to derive encryption key")?;
+
+        Ok(enc_key)
+    }
+
+    pub fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    pub fn sk_salt(&self) -> [u8; 32] {
+        self.sk_salt
+    }
+
+    pub fn address(&self) -> Address {
+        Address::p2wpkh(
+            &elements::bitcoin::PublicKey {
+                compressed: false,
+                key: self.get_public_key(),
+            },
+            Some(self.blinding_public_key()),
+            self.chain.into(),
+        )
+    }
+
+    fn blinding_public_key(&self) -> PublicKey {
+        PublicKey::from_secret_key(SECP256K1, &self.blinding_secret_key())
+    }
+
+    /// Derive the blinding key.
+    ///
+    /// # Choice of salt
+    ///
+    /// We choose to not add a salt because the ikm is already a randomly-generated, secret value with decent entropy.
+    ///
+    /// # Choice of ikm
+    ///
+    /// We derive the blinding key from the secret key to avoid having to store two secret values on disk.
+    ///
+    /// # Choice of info
+    ///
+    /// We choose to tag the derived key with `b"BLINDING_KEY"` in case we ever want to derive something else from the secret key.
+    pub fn blinding_secret_key(&self) -> SecretKey {
```

Comment:
Can we make this private?

Reply:
Used in a couple places still: https://github.com/comit-network/waves/blob/extract-wallet/extension/wallet/src/protocol/extract_trade.rs#L42.

But it's a good point. I think we should strive to not expose this one. Coin selection can just return the blinding key corresponding to each output it chooses, for example.